pkg/server/commands/listusers/list_users_rpc.go

```diff
@@ -324,38 +318,70 @@ func (l *listUsersQuery) expandIntersection(
 	pool.WithCancelOnError()
 	pool.WithMaxGoroutines(int(l.resolveNodeBreadthLimit))
 
-	intersectionFoundUsersChan := make(chan *openfgav1.User, 1)
+	wildcardKey := fmt.Sprintf("%s:*", req.GetUserFilters()[0].GetType())
 
-	children := rewrite.Intersection.GetChild()
-	for _, childRewrite := range children {
-		copyChildRewrite := childRewrite
+	clauses := rewrite.Intersection.GetChild()
+	intersectionFoundUsersChans := make([]chan *openfgav1.User, len(clauses))
+	for i, rewrite := range clauses {
+		i := i
+		rewrite := rewrite
+		intersectionFoundUsersChans[i] = make(chan *openfgav1.User, 1)
 		pool.Go(func(ctx context.Context) error {
-			return l.expandRewrite(ctx, req, copyChildRewrite, intersectionFoundUsersChan)
+			return l.expandRewrite(ctx, req, rewrite, intersectionFoundUsersChans[i])
 		})
 	}
 
 	errChan := make(chan error, 1)
 	go func() {
 		err := pool.Wait()
-		close(intersectionFoundUsersChan)
+		for i := range intersectionFoundUsersChans {
+			close(intersectionFoundUsersChans[i])
+		}
 		errChan <- err
 		close(errChan)
 	}()
 
+	var mu sync.Mutex
+	var wg sync.WaitGroup
+	wg.Add(len(clauses))
+
+	wildcardCount := atomic.Uint32{}
 	foundUsersCountMap := make(map[string]uint32, 0)
-	for fu := range intersectionFoundUsersChan {
-		key := tuple.UserProtoToString(fu)
-		foundUsersCountMap[key]++
+	for _, foundUsersChan := range intersectionFoundUsersChans {
+		go func(channel chan *openfgav1.User) {
+			foundUsersMap := make(map[string]uint32, 0)
+			for fu := range channel {
+				key := tuple.UserProtoToString(fu)
+				foundUsersMap[key]++
+			}
+
+			_, wildcardExists := foundUsersMap[wildcardKey]
+			if wildcardExists {
+				wildcardCount.Add(1)
+			}
+			for userKey := range foundUsersMap {
+				mu.Lock()
+				foundUsersCountMap[userKey]++
+				if wildcardExists {
+					foundUsersCountMap[userKey]--
+				}
+				mu.Unlock()
+			}
+			wg.Done()
+		}(foundUsersChan)
 	}
+	wg.Wait()
 
-	for key, c := range foundUsersCountMap {
-		// Compare the specific user's count, or the number of times
-		// the user was returned for all intersection clauses.
-		// If this count equals the number of clauses, the user satisfies
+	for key, count := range foundUsersCountMap {
+		// Compare the number of times the specific user was returned for
+		// all intersection clauses plus the number of wildcards.
+		// If this summed value equals the number of clauses, the user satisfies
 		// the intersection expression and can be sent on `foundUsersChan`
-		if c == uint32(len(children)) {
-			foundUsersChan <- tuple.StringToUserProto(key)
+		if (count + wildcardCount.Load()) < uint32(len(clauses)) {
+			continue
 		}
+
+		foundUsersChan <- tuple.StringToUserProto(key)
 	}
```

Comment:
I'm not a huge fan of this approach. It probably works but it looks complex, and you'll have to repeat it for exclusion. 

I feel like we're making our lives harder by not re-using the ReverseExpansion approach of using the graph traversal 🤔

Reply:
> I'm not a huge fan of this approach. It probably works but it looks complex, and you'll have to repeat it for exclusion.

I'm not in love either but it works and it makes sense why it works, although this is hugely subjective. 

Fortunately we don't need to repeat for exclusion. The complexity stems from supporting any number of clauses whereas exclusion always deal with two clauses.

> I feel like we're making our lives harder by not re-using the ReverseExpansion approach of using the graph traversal 

I still intend on having us evaluate wether reverse expand and/or some other algorithm is more appropriate, but for now I'm mostly concerned on getting an initial feature-complete version implemented.

Reply:
I don’t have any strong opposition to the algorithm at this time. It seems reasonable for the current API invariants. It would get much uglier if we supported more than 1 user filter in the near future. That would be my main concern with the approach, because that multiply the number of typed wildcards and user objects we'd have to track independently as well.

Reply:
I also don't have a strong opposition if it works, so i'm not blocking the PR :) 

 I'm just concerned about the fact that we have two very different implementations of two APIs that do relatively similar things, with all the downsides that that implies (maintainability, cognitive overload etc)